### PR TITLE
add boft support in stable-diffusion

### PR DESCRIPTION
**Pull request by sywangyi**
# What does this PR do?

Fixes # (issue)

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [ ] Did you make sure to update the documentation with your changes?
- [ ] Did you write any new necessary tests?

### CODE REVIEW

**Comment by imangohari1** [comment]
Hi @sywangyi 
Thanks for this PR. 
Could you rebase this with OH main and make sure `make style` is applied. 
Please share the results of the CI tests for test_diffusers.py with and without this changes. 
Thanks.

**Comment by sywangyi** [comment]
same with latest main. 1 case fail

FAILED tests/test_diffusers.py::GaudiStableDiffusionXLImg2ImgPipelineTests::test_stable_diffusion_xl_img2img_euler - AssertionError: 0.21911774845123289 not less than 0.01
================================ 1 failed, 108 passed, 46 skipped, 274 warnings in 776.40s (0:12:56) =================================

**Review by imangohari1 — CHANGES_REQUESTED** [review]
Hi @sywangyi 
I spent some time on this PR and did some testing:
- I've reworked the README file for a better read. Please apply the changes with the attached patch using `git am < 000*` (**don't copy past the changes, apply the patch please**). 
[0001-fea-dreambooth-reworked-the-readme.patch](https://github.com/user-attachments/files/16999762/0001-fea-dreambooth-reworked-the-readme.patch)

- I've tested the [PEFT example](https://github.com/huggingface/optimum-habana/tree/main/examples/stable-diffusion/training#peft-model-finetune) with both `lora` and `boft`. The lora example finishes in about 6min (`5m47.993s`) but the `boft` one has been running for ~80min and only compeleted 24% (`Steps:  24%|██▎       | 188/800 [1:21:18<3:53:56, 22.94s/it, loss=0.0225, lr=0.0001]`). Any thoughts on why `boft` is so significantly slower than `lora`? Is this bc of lack of hpu graphs? Let's investigate this a bit more. 
   - I've provided the tested `cmd` below. 

```
export MODEL_NAME="CompVis/stable-diffusion-v1-4"
export INSTANCE_DIR="dog"
export CLASS_DIR="path-to-class-images"
export OUTPUT_DIR="out"

logfile=pr1295.$(date -u +%Y%m%d%H%M).$(hostname).log

time python ../../gaudi_spawn.py --world_size 8 --use_mpi train_dreambooth.py   --pretrained_model_name_or_path=$MODEL_NAME    --instance_data_dir=$INSTANCE_DIR   --output_dir=$OUTPUT_DIR   --class_data_dir=$CLASS_DIR   --with_prior_preservation --prior_loss_weight=1.0   --instance_prompt="a photo of sks dog"   --class_prompt="a photo of dog"   --resolution=512   --train_batch_size=1   --num_class_images=200   --gradient_accumulation_steps=1   --learning_rate=1e-4   --lr_scheduler="constant"   --lr_warmup_steps=0   --max_train_steps=800   --mixed_precision=bf16   --use_hpu_graphs_for_training   --use_hpu_graphs_for_inference   --gaudi_config_name Habana/stable-diffusion   lora --unet_r 8 --unet_alpha 8 2>&1 | tee $logfile

time python ../../gaudi_spawn.py --world_size 8 --use_mpi train_dreambooth.py   --pretrained_model_name_or_path=$MODEL_NAME    --instance_data_dir=$INSTANCE_DIR   --output_dir=$OUTPUT_DIR   --class_data_dir=$CLASS_DIR   --with_prior_preservation --prior_loss_weight=1.0   --instance_prompt="a photo of sks dog"   --class_prompt="a photo of dog"   --resolution=512   --train_batch_size=1   --num_class_images=200   --gradient_accumulation_steps=1   --learning_rate=1e-4   --lr_scheduler="constant"   --lr_warmup_steps=0   --max_train_steps=800   --mixed_precision=bf16     --gaudi_config_name Habana/stable-diffusion   boft 2>&1 | tee $logfile
```

**Comment by sywangyi** [comment]
> Hi @sywangyi I spent some time on this PR and did some testing:
> 
> * I've reworked the README file for a better read. Please apply the changes with the attached patch using `git am < 000*` (**don't copy past the changes, apply the patch please**).
>   [0001-fea-dreambooth-reworked-the-readme.patch](https://github.com/user-attachments/files/16999762/0001-fea-dreambooth-reworked-the-readme.patch)
> * I've tested the [PEFT example](https://github.com/huggingface/optimum-habana/tree/main/examples/stable-diffusion/training#peft-model-finetune) with both `lora` and `boft`. The lora example finishes in about 6min (`5m47.993s`) but the `boft` one has been running for ~80min and only compeleted 24% (`Steps:  24%|██▎       | 188/800 [1:21:18<3:53:56, 22.94s/it, loss=0.0225, lr=0.0001]`). Any thoughts on why `boft` is so significantly slower than `lora`? Is this bc of lack of hpu graphs? Let's investigate this a bit more.
>   
>   * I've provided the tested `cmd` below.
> 
> ```
> export MODEL_NAME="CompVis/stable-diffusion-v1-4"
> export INSTANCE_DIR="dog"
> export CLASS_DIR="path-to-class-images"
> export OUTPUT_DIR="out"
> 
> logfile=pr1295.$(date -u +%Y%m%d%H%M).$(hostname).log
> 
> time python ../../gaudi_spawn.py --world_size 8 --use_mpi train_dreambooth.py   --pretrained_model_name_or_path=$MODEL_NAME    --instance_data_dir=$INSTANCE_DIR   --output_dir=$OUTPUT_DIR   --class_data_dir=$CLASS_DIR   --with_prior_preservation --prior_loss_weight=1.0   --instance_prompt="a photo of sks dog"   --class_prompt="a photo of dog"   --resolution=512   --train_batch_size=1   --num_class_images=200   --gradient_accumulation_steps=1   --learning_rate=1e-4   --lr_scheduler="constant"   --lr_warmup_steps=0   --max_train_steps=800   --mixed_precision=bf16   --use_hpu_graphs_for_training   --use_hpu_graphs_for_inference   --gaudi_config_name Habana/stable-diffusion   lora --unet_r 8 --unet_alpha 8 2>&1 | tee $logfile
> 
> time python ../../gaudi_spawn.py --world_size 8 --use_mpi train_dreambooth.py   --pretrained_model_name_or_path=$MODEL_NAME    --instance_data_dir=$INSTANCE_DIR   --output_dir=$OUTPUT_DIR   --class_data_dir=$CLASS_DIR   --with_prior_preservation --prior_loss_weight=1.0   --instance_prompt="a photo of sks dog"   --class_prompt="a photo of dog"   --resolution=512   --train_batch_size=1   --num_class_images=200   --gradient_accumulation_steps=1   --learning_rate=1e-4   --lr_scheduler="constant"   --lr_warmup_steps=0   --max_train_steps=800   --mixed_precision=bf16     --gaudi_config_name Habana/stable-diffusion   boft 2>&1 | tee $logfile
> ```

yes. I have file a bug to pytorch training team about the perf issue, will cc you in the jira

**Comment by imangohari1** [comment]
Overal LGTM although there is a performance issue with boft. Thanks @sywangyi 
@regisss how do you suggest we proceed?

**Comment by libinta** [comment]
@sywangyi  please test with latest synapse SW, if there is still issue, we dont need to merge this change for next synapse release as it's not functional

**Comment by sywangyi** [comment]
> @sywangyi please test with latest synapse SW, if there is still issue, we dont need to merge this change for next synapse release as it's not functional

which version do you mean, I think habana pytorch training team is still working on it.

**Comment by imangohari1** [comment]
> > @sywangyi please test with latest synapse SW, if there is still issue, we dont need to merge this change for next synapse release as it's not functional
> 
> which version do you mean? I think habana pytorch training team is still working on it.

@sywangyi  1.18.0 release build id 410.

**Comment by libinta** [comment]
@sywangyi  do you have test result?

**Comment by kaixuanliu** [comment]
> @sywangyi do you have test result?

@libinta , Hi, we do not have env for 1.18.0 with build 410, but @yeonsily gave me a docker image with build 438, can we use it to have a test? The env for this docker image is:
```
habana-torch-dataloader 1.18.0.438
habana-torch-plugin     1.18.0.438
pytorch-lightning       2.4.0
torch                   2.4.0a0+git74cd574
torch_tb_profiler       0.4.0
torchaudio              2.4.0a0+69d4077
torchdata               0.7.1+5e6f7b7
torchmetrics            1.4.2
torchsde                0.2.6
torchtext               0.18.0a0+9bed85d
torchvision             0.19.0a0+48b1edf
```

**Comment by kaixuanliu** [comment]
> @sywangyi do you have test result?

Hi, @libinta , we used a docker image with build 438 for 1.18.0 release from @yeonsily and have a test. The perf of boft is still significantly slower than lora, for single card finetune, lora takes ~4min 30s, while boft needs more than 2 hours.

**Comment by imangohari1** [comment]
> > @sywangyi do you have test result?
> 
> Hi, @libinta , we used a docker image with build 438 for 1.18.0 release from @yeonsily and have a test. The perf of boft is still significantly slower than lora, for single card finetune, lora takes ~4min 30s, while boft needs more than 2 hours.

I have tested this with driver 1.18.0-460 and the corresponding docker. It still shows the same behavior. 
Our team is looking at rewriting the code that is causing recompile.

**Comment by yao-matrix** [comment]
> > > @sywangyi do you have test result?
> > 
> > 
> > Hi, @libinta , we used a docker image with build 438 for 1.18.0 release from @yeonsily and have a test. The perf of boft is still significantly slower than lora, for single card finetune, lora takes ~4min 30s, while boft needs more than 2 hours.
> 
> I have tested this with driver 1.18.0-460 and the corresponding docker. It still shows the same behavior. Our team is looking at rewriting the code that is causing recompile.

@imangohari1 , do we have update on this?

**Comment by imangohari1** [comment]
> > > > @sywangyi do you have test result?
> > > 
> > > 
> > > Hi, @libinta , we used a docker image with build 438 for 1.18.0 release from @yeonsily and have a test. The perf of boft is still significantly slower than lora, for single card finetune, lora takes ~4min 30s, while boft needs more than 2 hours.
> > 
> > 
> > I have tested this with driver 1.18.0-460 and the corresponding docker. It still shows the same behavior. Our team is looking at rewriting the code that is causing recompile.
> 
> @imangohari1 , do we have update on this?

I am not if the issue is resolved or not. 
@sywangyi WDYT?

**Comment by sywangyi** [comment]
according to https://habana.atlassian.net/browse/HS-3208, it has not been resolved yet

**Comment by Luca-Calabria** [comment]
Update: RnD guys said the new feature is targeted for release 1.21.0

**Comment by Luca-Calabria** [comment]
The fix is under PR review, I'll test it when it is ready

**Comment by Luca-Calabria** [comment]
@sywangyi the build with the fix is ready 1.21.0-376. I tested it. Please check the ticket for all the details.

**Comment by Luca-Calabria** [comment]
@sywangyi let me try again. In my test removing .cpu() from all the three orth_rotate_butterfly.squeeze(0).cpu() and running lazy mode I registrered 7 minutes to complete 1 epoch with 10 classes. 
Anyway the build 1.21.376 need explicit PT_HPU_LAZY_MODE=1 to run Lazy

**Comment by Luca-Calabria** [comment]
@sywangyi I'm not sure what is happening in your case but I'm able to run multiple epochs
image

**Comment by sywangyi** [comment]
yes, 32min in my env, is the perf expected?

**Comment by sywangyi** [comment]
@libinta  will we merge the PR in 1.21, since 1.21 fix the the slow compilation when "torch.block_diag" operation run

**Comment by sywangyi** [comment]
> Does this PR work with Synapse 1.21?

yes

File: tests/test_diffusers.py [hunk 1]
```diff
@@ -615,6 +697,45 @@ def test_stable_diffusion_hpu_graphs(self):
...
+            use_hpu_graphs=True,
+            gaudi_config=GaudiConfig.from_pretrained("Habana/stable-diffusion"),
+            torch_dtype=torch.bfloat16,
+        )
+        if peft_adapter not in ["boft", "oft"]:
+            with torch.autocast(device_type="hpu", dtype=torch.bfloat16, enabled=True):
+                pipeline.unet = self.merge_peft_adapter(pipeline.unet, peft_adapter)
+                pipeline.text_encoder = self.merge_peft_adapter(pipeline.text_encoder, peft_adapter)
+        else:
+            # WA torch.inverse issue in Synapse AI 1.17 for oft and boft
```
Comment (regisss) [inline tests/test_diffusers.py]: Do we still need this?

File: examples/stable-diffusion/text_to_image_generation.py [hunk 2]
```diff
@@ -576,19 +576,50 @@ def main():
...
+                    from optimum.habana.peft.layer import GaudiBoftGetDeltaWeight
+
+                    tuners.boft.layer.Linear.get_delta_weight = GaudiBoftGetDeltaWeight
+                    tuners.boft.layer.Conv2d.get_delta_weight = GaudiBoftGetDeltaWeight
+                    tuners.boft.layer._FBD_CUDA = False
 
                     pipeline.unet = PeftModel.from_pretrained(pipeline.unet, args.unet_adapter_name_or_path)
-                    pipeline.unet = pipeline.unet.merge_and_unload()
+                    if pipeline.unet.peft_type in [PeftType.OFT, PeftType.BOFT]:
+                        # WA torch.inverse issue in Synapse AI 1.17 for oft and boft
```
Comment (regisss) [inline examples/stable-diffusion/text_to_image_generation.py]: Do we still need this?

File: examples/stable-diffusion/text_to_image_generation.py [hunk 3]
```diff
@@ -576,19 +576,50 @@ def main():
...
+                    tuners.boft.layer.Conv2d.get_delta_weight = GaudiBoftGetDeltaWeight
+                    tuners.boft.layer._FBD_CUDA = False
 
                     pipeline.text_encoder = PeftModel.from_pretrained(
                         pipeline.text_encoder, args.text_encoder_adapter_name_or_path
                     )
-                    pipeline.text_encoder = pipeline.text_encoder.merge_and_unload()
-
+                    if pipeline.text_encoder.peft_type in [PeftType.OFT, PeftType.BOFT]:
+                        # WA torch.inverse issue in Synapse AI 1.17 for oft and boft
```
Comment (regisss) [inline examples/stable-diffusion/text_to_image_generation.py]: Same

File: examples/stable-diffusion/text_to_image_generation.py [hunk 4]
```diff
@@ -576,19 +576,24 @@ def main():
...
                 if args.unet_adapter_name_or_path is not None:
-                    from peft import PeftModel
+                    from peft import PeftModel, tuners
+
+                    tuners.boft.layer._FBD_CUDA = False
 
                     pipeline.unet = PeftModel.from_pretrained(pipeline.unet, args.unet_adapter_name_or_path)
-                    pipeline.unet = pipeline.unet.merge_and_unload()
+                    with torch.autocast(device_type="hpu", dtype=torch.bfloat16, enabled=args.bf16):
+                        pipeline.unet = pipeline.unet.merge_and_unload()
```
Comment (regisss) [inline examples/stable-diffusion/text_to_image_generation.py:585]: I'm curious, autocast has an impact on merge_and_unload ?

**Review by regisss — COMMENTED** [review]
It looks good to me. Can you also check that the Diffusers CI passes please? CI is offline and I don't have access to any Gaudi server at the moment.

File: examples/stable-diffusion/text_to_image_generation.py (hunk 4 above)
Comment (sywangyi) [reply examples/stable-diffusion/text_to_image_generation.py:585]: has impact on oft in lazy model, remove the autocast will lead to "RuntimeError: Graph compile failed. synStatus=synStatus 26 [Generic failure]"
